File: src/main/java/net/sf/rails/game/model/ShareDetailsModel.java
package net.sf.rails.game.model;

import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.financial.PublicCertificate;

import com.google.common.collect.Multimap;


/**
 * Model for displaying the share details (used for ToolTips)
*/
public final class ShareDetailsModel extends RailsModel {
        
    private final PublicCompany company;

    private ShareDetailsModel(CertificatesModel parent, PublicCompany company) {
        super(parent, "shareDetailsModel_" + company.getId());
        this.company = company;
    }

    public static ShareDetailsModel create(CertificatesModel certModel, PublicCompany company) {
        ShareDetailsModel model = new ShareDetailsModel(certModel, company);
        certModel.addModel(model);
        return model;
    }
    
    @Override
    public CertificatesModel getParent() {
        return (CertificatesModel)super.getParent();
    }
    
    @Override
    public String toText() {
        Multimap<String, PublicCertificate> certs = getParent().getCertificatesByType(company);
        if (certs.isEmpty()) return null;

        StringBuilder text = new StringBuilder();
        for (String certType : certs.keySet()) {
            if (text.length() > 0) text.append("<br>");
            // parse certType
            // TODO: Create true CertificateTypes
            String[] items = certType.split("_");
            String type = items[1] + (items.length > 2 && items[2].contains("P") ? "P" : "");
            text.append(type).append(" x ").append(certs.get(certType).size());
        }
        return "<html>" + text.toString() + "</html>";
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/ShareDetailsModel.cs
using GameLib.Net.Game.Financial;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * Model for displaying the share details (used for ToolTips)
*/

namespace GameLib.Net.Game.Model
{
    public class ShareDetailsModel : RailsModel
    {
        private PublicCompany company;

    private ShareDetailsModel(CertificatesModel parent, PublicCompany company) : base(parent, "shareDetailsModel_" + company.Id)
        {
            
            this.company = company;
        }

        public static ShareDetailsModel Create(CertificatesModel certModel, PublicCompany company)
        {
            ShareDetailsModel model = new ShareDetailsModel(certModel, company);
            certModel.AddModel(model);
            return model;
        }

        new public CertificatesModel Parent
        {
            get
            {
                return (CertificatesModel)base.Parent;
            }
        }

    override public string ToText()
        {
            Wintellect.PowerCollections.MultiDictionary<string, PublicCertificate> certs = Parent.GetCertificatesByType(company);
            if (certs.Count == 0) return null;

            StringBuilder text = new StringBuilder();
            foreach (string certType in certs.Keys)
            {
                if (text.Length > 0) text.Append("<br>");
                // parse certType
                // TODO: Create true CertificateTypes
                string[] items = certType.Split('_');
                string type = items[1] + (items.Length > 2 && items[2].Contains("P") ? "P" : "");
                text.Append(type).Append(" x ").Append(certs[certType].Count);
            }
            return "<html>" + text.ToString() + "</html>";
        }
    }
}
